daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.daffodil.infoset
+
+import java.lang.{ Boolean => JBoolean, Number => JNumber, Long => JLong, Double => JDouble, String => JString, Float => JFloat, Byte => JByte, Integer => JInt, Short => JShort }
+import java.math.{ BigDecimal => JBigDecimal, BigInteger => JBigInt }
+import org.apache.daffodil.calendar.DFDLCalendar
+import scala.language.implicitConversions
+import org.apache.daffodil.calendar.DFDLDateTime
+import org.apache.daffodil.calendar.DFDLDate
+import org.apache.daffodil.calendar.DFDLTime
+import org.apache.daffodil.util.Maybe
+import org.apache.daffodil.processors.UseNilForDefault
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.util.Maybe.One
+import org.apache.daffodil.util.Maybe.Nope
+import java.io.ObjectOutputStream
+import java.io.ObjectInputStream
+import java.net.URI
+
+/*
+ * These traits are used for the phantom type X. When considering the type of a
+ * DataValue. See the comment on the DataValue class below.
+ */
+trait NonNullable extends Nullable;
+trait Nullable extends tUseNilForDefault;
+trait tUseNilForDefault;
+sealed trait DataValuePrimitiveType
+
+/*
+ * The type variable X here is a phantom type. That is to say, it is not actually
+ * used to describe the type of anything within DataValue. The goal is to provide us
+ * with more ability to control the inheritense relationship between various DataValue types.
+ * In theory, this would be accomplished by establishing a type hierarchy on the types
+ * used for T. Indeed, if Scala had type-classes as a first-class feature that is likely
+ * the approach we would have taken. However, since most of the types we use for T
+ * are out of our control, we are unable to control what type structure they have.
+ * By carefully selecting X in our type aliases, it is possible for to define types
+ * that can represent any subset of types which we want. It is further possible to
+ * allow for automatic upcasting between these types where appropriate.
+ *
+ * We inline all methods because attempts to call non-inlined methods trigger boxing
```

Comment:
I don't think that's right. Did you disassemble the byte code? I think polymorphic usage of AnyVal types triggers boxing, but just calling a method on one doesn't. It's the equivalent of calling a Java static method.

Reply:
You are correct. these methods do not trigger boxxing.

Also, the inline annotation appears to do nothing. When I tested, AnyVal methods get compiled and used as static methods regardless of the inline annotation.

Reply:
Perhaps doesn't inline because we have debug flags in our compile options?

Reply:
I removed the inline portion of this comment.

I did my testing using the scalac directly from the commandline. It wouldn't suprise me if debug mode was the default though.

Regardless, I see no reason not to inline these.